\chapter{A scalable interface between data and analysis underneath R}
\hypertarget{index}{}\label{index}\index{A scalable interface between data and analysis underneath R@{A scalable interface between data and analysis underneath R}}
  Reading genomic data files (\href{https://www.ebi.ac.uk/training/online/courses/human-genetic-variation-introduction/variant-identification-and-analysis/understanding-vcf-format/}{\texttt{ VCF}}, \href{https://samtools.github.io/bcftools/howtos/index.html}{\texttt{ BCF}}, \href{https://www.chg.ox.ac.uk/~gav/bgen_format/index.html}{\texttt{ BGEN}}, \href{https://anndata.readthedocs.io/en/latest/index.html}{\texttt{ H5\+AD}}, \href{https://bioconductor.org/packages/DelayedArray}{\texttt{ Delayed\+Array}}) into R/\+Rcpp in chunks for analysis with \href{https://doi.org/10.21105/joss.00026}{\texttt{ Armadillo}} / \href{eigen.tuxfamily.org}{\texttt{ Eigen}} / \href{https://www.rcpp.org}{\texttt{ Rcpp}} libraries. Mondern datasets are often too big to fit into memory, and many analyses operate a small chunk features at a time. Yet in practice, many implementations require the whole dataset stored in memory. Others pair an analysis with a specific data format (i.\+e. regresson analysis paired with genotype data from a VCF) in way that the two components can\textquotesingle{}t be separated for use in other applications.  The {\ttfamily Genomic\+Data\+Stream} C++ inferface separates


\begin{DoxyEnumerate}
\item data source
\item streaming chunks of features into a data matrix
\item downstream analysis ~\newline

\end{DoxyEnumerate}\hypertarget{index_autotoc_md0}{}\doxysubsection{\texorpdfstring{Dependencies}{Dependencies}}\label{index_autotoc_md0}
\tabulinesep=1mm
\begin{longtabu}spread 0pt [c]{*{3}{|X[-1]}|}
\hline
\PBS\centering \cellcolor{\tableheadbgcolor}\textbf{ Package   }&\PBS\centering \cellcolor{\tableheadbgcolor}\textbf{ Ref   }&\PBS\centering \cellcolor{\tableheadbgcolor}\textbf{ Role    }\\\cline{1-3}
\endfirsthead
\hline
\endfoot
\hline
\PBS\centering \cellcolor{\tableheadbgcolor}\textbf{ Package   }&\PBS\centering \cellcolor{\tableheadbgcolor}\textbf{ Ref   }&\PBS\centering \cellcolor{\tableheadbgcolor}\textbf{ Role    }\\\cline{1-3}
\endhead
\href{https://cran.r-project.org/package=vcfppR}{\texttt{ vcfppR}}   &\href{https://doi.org/10.1093/bioinformatics/btae049}{\texttt{ Bioinformatics}}   &C++ API for htslib    \\\cline{1-3}
\href{https://github.com/samtools/htslib}{\texttt{ htslib}}   &\href{https://doi.org/10.1093/gigascience/giab007}{\texttt{ Giga\+Science}}   &C API for VCF/\+BCF files    \\\cline{1-3}
\href{https://bioconductor.org/packages/beachmat/}{\texttt{ beatchmat}}   &\href{https://doi.org/10.1371/journal.pcbi.1006135}{\texttt{ PLoS Comp Biol}}   &C++ API for access data owned by R    \\\cline{1-3}
\href{https://cran.r-project.org/package=Rcpp}{\texttt{ Rcpp}}   &\href{https://doi.org/10.18637/jss.v040.i08}{\texttt{ J Stat Software}}   &API for R/\+C++ integration    \\\cline{1-3}
\href{https://cran.r-project.org/package=RcppEigen}{\texttt{ Rcpp\+Eigen}}   &\href{https://doi.org/10.18637/jss.v052.i05}{\texttt{ J Stat Software}}   &API for Rcpp access to Eigen matrix library    \\\cline{1-3}
\href{https://cran.r-project.org/package=RcppArmadillo}{\texttt{ Rcpp\+Armadillo}}   &\href{https://doi.org/10.18637/jss.v040.i08}{\texttt{ J Stat Software}}   &API for Rcpp access to Armadillo matrix library    \\\cline{1-3}
\href{eigen.tuxfamily.org}{\texttt{ Eigen}}   &&C++ library for linear algebra with advanced features    \\\cline{1-3}
\href{https://arma.sourceforge.net}{\texttt{ Armadillo}}   &\href{https://doi.org/10.21105/joss.00026}{\texttt{ J Open Src Soft}}   &User-\/friendly C++ library for linear algebra   \\\cline{1-3}
\end{longtabu}
