<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.12.0"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<!-- BEGIN opengraph metadata -->
<meta property="og:title" content="Doxygen Awesome" />
<meta property="og:image" content="https://repository-images.githubusercontent.com/348492097/4f16df80-88fb-11eb-9d31-4015ff22c452" />
<meta property="og:description" content="Custom CSS theme for doxygen html-documentation with lots of customization parameters." />
<meta property="og:url" content="https://jothepro.github.io/doxygen-awesome-css/" />
<!-- END opengraph metadata -->
<!-- BEGIN twitter metadata -->
<meta name="twitter:image:src" content="https://repository-images.githubusercontent.com/348492097/4f16df80-88fb-11eb-9d31-4015ff22c452" />
<meta name="twitter:title" content="Doxygen Awesome" />
<meta name="twitter:description" content="Custom CSS theme for doxygen html-documentation with lots of customization parameters." />
<!-- END twitter metadata -->
<title>GenomicDataStream: A scalable interface between data and analysis underneath R</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<link rel="icon" type="image/svg+xml" href="logo.drawio.svg"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<script type="text/javascript" src="doxygen-awesome-darkmode-toggle.js"></script>
<script type="text/javascript" src="doxygen-awesome-fragment-copy-button.js"></script>
<script type="text/javascript" src="doxygen-awesome-paragraph-link.js"></script>
<script type="text/javascript" src="doxygen-awesome-interactive-toc.js"></script>
<script type="text/javascript" src="doxygen-awesome-tabs.js"></script>
<script type="text/javascript" src="toggle-alternative-theme.js"></script>
<script type="text/javascript">
    DoxygenAwesomeFragmentCopyButton.init()
    DoxygenAwesomeDarkModeToggle.init()
    DoxygenAwesomeParagraphLink.init()
    DoxygenAwesomeInteractiveToc.init()
    DoxygenAwesomeTabs.init()
</script>
<link href="navtree.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="resize.js"></script>
<script type="text/javascript" src="cookie.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
<link href="doxygen-awesome.css" rel="stylesheet" type="text/css"/>
<link href="custom.css" rel="stylesheet" type="text/css"/>
</head>
<body>
<!-- https://tholman.com/github-corners/ -->
<a href="https://github.com/jothepro/doxygen-awesome-css" class="github-corner" title="View source on GitHub" target="_blank" rel="noopener noreferrer">
    <svg viewBox="0 0 250 250" width="40" height="40" style="position: absolute; top: 0; border: 0; right: 0; z-index: 99;" aria-hidden="true">
    <path d="M0,0 L115,115 L130,115 L142,142 L250,250 L250,0 Z"></path><path d="M128.3,109.0 C113.8,99.7 119.0,89.6 119.0,89.6 C122.0,82.7 120.5,78.6 120.5,78.6 C119.2,72.0 123.4,76.3 123.4,76.3 C127.3,80.9 125.5,87.3 125.5,87.3 C122.9,97.6 130.6,101.9 134.4,103.2" fill="currentColor" style="transform-origin: 130px 106px;" class="octo-arm"></path><path d="M115.0,115.0 C114.9,115.1 118.7,116.5 119.8,115.4 L133.7,101.6 C136.9,99.2 139.9,98.4 142.2,98.6 C133.8,88.0 127.5,74.4 143.8,58.0 C148.5,53.4 154.0,51.2 159.7,51.0 C160.3,49.4 163.2,43.6 171.4,40.1 C171.4,40.1 176.1,42.5 178.8,56.2 C183.1,58.6 187.2,61.8 190.9,65.4 C194.5,69.0 197.7,73.2 200.1,77.6 C213.8,80.2 216.3,84.9 216.3,84.9 C212.7,93.1 206.9,96.0 205.4,96.6 C205.1,102.4 203.0,107.8 198.3,112.5 C181.9,128.9 168.3,122.5 157.7,114.1 C157.9,116.9 156.7,120.9 152.7,124.9 L141.0,136.5 C139.8,137.7 141.6,141.9 141.8,141.8 Z" fill="currentColor" class="octo-body"></path></svg></a><style>.github-corner:hover .octo-arm{animation:octocat-wave 560ms ease-in-out}@keyframes octocat-wave{0%,100%{transform:rotate(0)}20%,60%{transform:rotate(-25deg)}40%,80%{transform:rotate(10deg)}}@media (max-width:500px){.github-corner:hover .octo-arm{animation:none}.github-corner .octo-arm{animation:octocat-wave 560ms ease-in-out}}</style>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td id="projectlogo"><img alt="Logo" src="GDS_logo_55.png"/></td>
  <td id="projectalign" style="padding-left: 0.5em;">
   <div id="projectname">GenomicDataStream
   </div>
   <div id="projectbrief">A scalable interface between data and analysis</div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.12.0 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
var searchBox = new SearchBox("searchBox", "search/",'.html');
/* @license-end */
</script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
$(function() { codefold.init(0); });
/* @license-end */
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
$(function() {
  initMenu('',true,false,'search.php','Search',false);
  $(function() { init_search(); });
});
/* @license-end */
</script>
<div id="main-nav"></div>
</div><!-- top -->
<div id="doc-content">
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
$(function(){ initResizable(false); });
/* @license-end */
</script>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<div id="MSearchResults">
<div class="SRPage">
<div id="SRIndex">
<div id="SRResults"></div>
<div class="SRStatus" id="Loading">Loading...</div>
<div class="SRStatus" id="Searching">Searching...</div>
<div class="SRStatus" id="NoMatches">No Matches</div>
</div>
</div>
</div>
</div>

<div><div class="header">
  <div class="headertitle"><div class="title">A scalable interface between data and analysis underneath R </div></div>
</div><!--header-->
<div class="contents">
<div class="textblock"><div class="image">
<img src="GenomicDataStream.png" alt=""/>
</div>
 <div style="text-align: justify"> Reading genomic data files (<a href="https://www.ebi.ac.uk/training/online/courses/human-genetic-variation-introduction/variant-identification-and-analysis/understanding-vcf-format/">VCF</a>, <a href="https://samtools.github.io/bcftools/howtos/index.html">BCF</a>, <a href="https://www.chg.ox.ac.uk/~gav/bgen_format/index.html">BGEN</a>, <a href="https://www.cog-genomics.org/plink/2.0/input#pgen">PGEN</a>, <a href="https://www.cog-genomics.org/plink/2.0/input#bed">BED</a>, <a href="https://anndata.readthedocs.io/en/latest/index.html">H5AD</a>, <a href="https://bioconductor.org/packages/DelayedArray">DelayedArray</a>) into R/Rcpp in chunks for analysis with <a href="https://doi.org/10.21105/joss.00026">Armadillo</a> / <a href="eigen.tuxfamily.org">Eigen</a> / <a href="https://www.rcpp.org">Rcpp</a> libraries. Mondern datasets are often too big to fit into memory, and many analyses operate a small chunk features at a time. Yet in practice, many implementations require the whole dataset stored in memory. Others pair an analysis with a specific data format (i.e. regresson analysis paired with genotype data from a VCF) in way that the two components can't be separated for use in other applications. </div><h4><a class="anchor" id="autotoc_md0"></a>
The <code>GenomicDataStream</code> C++ interface separates</h4>
<ol type="1">
<li>data source</li>
<li>streaming chunks of features into a data matrix</li>
<li>downstream analysis <br  />
</li>
</ol>
<h3><a class="anchor" id="autotoc_md1"></a>
Example code with C++17</h3>
<div class="fragment"><div class="line"><span class="preprocessor">#include &lt;RcppArmadillo.h&gt;</span></div>
<div class="line"><span class="preprocessor">#include &lt;<a class="code" href="_genomic_data_stream_8h.html">GenomicDataStream.h</a>&gt;</span></div>
<div class="line"> </div>
<div class="line"><span class="comment">// use namespace for GenomicDataStream</span></div>
<div class="line"><span class="keyword">using namespace </span><a class="code hl_namespace" href="namespacegds.html">gds</a>;</div>
<div class="line"> </div>
<div class="line"><span class="comment">// parameters </span></div>
<div class="line"><span class="keywordtype">string</span> file = <span class="stringliteral">&quot;test.vcf.gz&quot;</span>;</div>
<div class="line"><span class="keywordtype">string</span> field = <span class="stringliteral">&quot;DS&quot;</span>;    <span class="comment">// read dosage field</span></div>
<div class="line"><span class="keywordtype">string</span> region = <span class="stringliteral">&quot;&quot;</span>;     <span class="comment">// no region filter</span></div>
<div class="line"><span class="keywordtype">string</span> samples = <span class="stringliteral">&quot;-&quot;</span>;   <span class="comment">// no samples filter</span></div>
<div class="line"><span class="keywordtype">int</span> chunkSize = 4;      <span class="comment">// each chunk will read 4 variants</span></div>
<div class="line"> </div>
<div class="line"><span class="comment">// initialize parameters</span></div>
<div class="line"><a class="code hl_struct" href="structgds_1_1_param.html">Param</a> param(file, region, samples, chunkSize);</div>
<div class="line">param.setField( field );</div>
<div class="line"> </div>
<div class="line"><span class="comment">// Initialise GenomicDataStream to read </span></div>
<div class="line"><span class="comment">// VCF/BCF and BGEN with same interface</span></div>
<div class="line">unique_ptr&lt;GenomicDataStream&gt; gdsStream = createFileView( param );</div>
<div class="line"> </div>
<div class="line"><span class="comment">// declare DataChunk storing an Armadillo matrix for each chunk</span></div>
<div class="line"><a class="code hl_class" href="classgds_1_1_data_chunk.html">DataChunk&lt;arma::mat&gt;</a> chunk;</div>
<div class="line"> </div>
<div class="line"><span class="comment">// Store meta-data about each variant</span></div>
<div class="line"><a class="code hl_class" href="classgds_1_1_variant_info.html">VariantInfo</a> *info;</div>
<div class="line"> </div>
<div class="line"><span class="comment">// loop through chunks</span></div>
<div class="line"><span class="keywordflow">while</span>( gdsStream-&gt;getNextChunk( chunk ) ){</div>
<div class="line"> </div>
<div class="line">    <span class="comment">// get data from chunk</span></div>
<div class="line">    <span class="comment">// chunk.getData();</span></div>
<div class="line"> </div>
<div class="line">    <span class="comment">// get variant information</span></div>
<div class="line">    info = chunk.<a class="code hl_function" href="classgds_1_1_data_chunk.html#ac16eb50ec9f3e0cf7d1ab46fb7a59531">getInfo</a>&lt;<a class="code hl_class" href="classgds_1_1_variant_info.html">VariantInfo</a>&gt;();</div>
<div class="line"> </div>
<div class="line">    <span class="comment">// Do analysis with variants in this chunk</span></div>
<div class="line">}</div>
<div class="ttc" id="a_genomic_data_stream_8h_html"><div class="ttname"><a href="_genomic_data_stream_8h.html">GenomicDataStream.h</a></div></div>
<div class="ttc" id="aclassgds_1_1_data_chunk_html"><div class="ttname"><a href="classgds_1_1_data_chunk.html">gds::DataChunk</a></div><div class="ttdef"><b>Definition</b> GenomicDataStream_virtual.h:35</div></div>
<div class="ttc" id="aclassgds_1_1_data_chunk_html_ac16eb50ec9f3e0cf7d1ab46fb7a59531"><div class="ttname"><a href="classgds_1_1_data_chunk.html#ac16eb50ec9f3e0cf7d1ab46fb7a59531">gds::DataChunk::getInfo</a></div><div class="ttdeci">infoType * getInfo()</div><div class="ttdef"><b>Definition</b> GenomicDataStream_virtual.h:52</div></div>
<div class="ttc" id="aclassgds_1_1_variant_info_html"><div class="ttname"><a href="classgds_1_1_variant_info.html">gds::VariantInfo</a></div><div class="ttdef"><b>Definition</b> VariantInfo.h:23</div></div>
<div class="ttc" id="anamespacegds_html"><div class="ttname"><a href="namespacegds.html">gds</a></div><div class="ttdef"><b>Definition</b> bgenstream.h:32</div></div>
<div class="ttc" id="astructgds_1_1_param_html"><div class="ttname"><a href="structgds_1_1_param.html">gds::Param</a></div><div class="ttdef"><b>Definition</b> GenomicDataStream_virtual.h:68</div></div>
</div><!-- fragment --><h2><a class="anchor" id="autotoc_md2"></a>
Dependencies</h2>
<table class="markdownTable">
<tr class="markdownTableHead">
<th class="markdownTableHeadNone">Package   </th><th class="markdownTableHeadNone">Ref   </th><th class="markdownTableHeadNone">Role    </th></tr>
<tr class="markdownTableRowOdd">
<td class="markdownTableBodyNone"><a href="https://cran.r-project.org/package=vcfppR">vcfppR</a>   </td><td class="markdownTableBodyNone"><a href="https://doi.org/10.1093/bioinformatics/btae049">Bioinformatics</a>   </td><td class="markdownTableBodyNone">C++ API for htslib    </td></tr>
<tr class="markdownTableRowEven">
<td class="markdownTableBodyNone"><a href="https://github.com/samtools/htslib">htslib</a>   </td><td class="markdownTableBodyNone"><a href="https://doi.org/10.1093/gigascience/giab007">GigaScience</a>   </td><td class="markdownTableBodyNone">C API for VCF/BCF files    </td></tr>
<tr class="markdownTableRowOdd">
<td class="markdownTableBodyNone"><a href="https://cran.r-project.org/package=pgenlibr">pgenlibr</a>   </td><td class="markdownTableBodyNone"><a href="https://doi.org/10.1186/s13742-015-0047-8">GigaScience</a>   </td><td class="markdownTableBodyNone">R/C++ API for plink files    </td></tr>
<tr class="markdownTableRowEven">
<td class="markdownTableBodyNone"><a href="https://bioconductor.org/packages/beachmat/">beatchmat</a>   </td><td class="markdownTableBodyNone"><a href="https://doi.org/10.1371/journal.pcbi.1006135">PLoS Comp Biol</a>   </td><td class="markdownTableBodyNone">C++ API for access data owned by R    </td></tr>
<tr class="markdownTableRowOdd">
<td class="markdownTableBodyNone"><a href="https://cran.r-project.org/package=Rcpp">Rcpp</a>   </td><td class="markdownTableBodyNone"><a href="https://doi.org/10.18637/jss.v040.i08">J Stat Software</a>   </td><td class="markdownTableBodyNone">API for R/C++ integration    </td></tr>
<tr class="markdownTableRowEven">
<td class="markdownTableBodyNone"><a href="https://cran.r-project.org/package=RcppEigen">RcppEigen</a>   </td><td class="markdownTableBodyNone"><a href="https://doi.org/10.18637/jss.v052.i05">J Stat Software</a>   </td><td class="markdownTableBodyNone">API for Rcpp access to Eigen matrix library    </td></tr>
<tr class="markdownTableRowOdd">
<td class="markdownTableBodyNone"><a href="https://cran.r-project.org/package=RcppArmadillo">RcppArmadillo</a>   </td><td class="markdownTableBodyNone"><a href="https://doi.org/10.18637/jss.v040.i08">J Stat Software</a>   </td><td class="markdownTableBodyNone">API for Rcpp access to Armadillo matrix library    </td></tr>
<tr class="markdownTableRowEven">
<td class="markdownTableBodyNone"><a href="https://eigen.tuxfamily.org">Eigen</a>   </td><td class="markdownTableBodyNone"></td><td class="markdownTableBodyNone">C++ library for linear algebra with advanced features    </td></tr>
<tr class="markdownTableRowOdd">
<td class="markdownTableBodyNone"><a href="https://arma.sourceforge.net">Armadillo</a>   </td><td class="markdownTableBodyNone"><a href="https://doi.org/10.21105/joss.00026">J Open Src Soft</a>   </td><td class="markdownTableBodyNone">User-friendly C++ library for linear algebra   </td></tr>
</table>
<h2><a class="anchor" id="autotoc_md3"></a>
Notes</h2>
<div style="text-align: justify"> <code>GenomicDataStream</code> provide flexability in terms of data input types and and matrix libraries. This can useful in many cases, but the large number of dependencies can require installation of additional libraries and increase compile times. Some of these dependencies can be avoided by removing support for some capabilities with compiler flags in <code>Makevars</code>: </div><p><code>-D DISABLE_DELAYED_STREAM</code> <br  />
 &#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160; Omit <code>DelayedStream</code> class, remove dependence on <code>Rcpp</code> and <code>beachmat</code> <br  />
</p>
<p><code>-D DISABLE_EIGEN</code> <br  />
 &#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160; Omit support for Eigen matrix library, and remove dependence on <code>RcppEigen</code> and <code>Eigen</code></p>
<p><code>-D DISABLE_RCPP</code> <br  />
 &#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160; Omit support for <code>Rcpp</code> matrix library, and remove dependence on <code>Rcpp</code></p>
<p><code>-D DISABLE_PLINK</code> <br  />
 &#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160; Omit support for <code>PLINK</code> files (PGEN, BED), and remove dependence on <code>pgenlibr</code></p>
<p><br  />
 <br  />
 <img src="https://labs.icahn.mssm.edu/msdw/wp-content/uploads/sites/350/2020/03/sinai_logo.jpg" alt="" align="left" width="40px/" class="inline"/> &#160;&#160;Developed by <a href="http://gabrielhoffman.github.io">Gabriel Hoffman</a> at <a href="https://icahn.mssm.edu/research/neurogenomics">Center for Disease Neurogenomics</a> at the <a href="https://icahn.mssm.edu">Icahn School of Medicine at Mount Sinai</a>. </p>
</div></div><!-- PageDoc -->
<a href="doxygen_crawl.html"></a>
</div><!-- contents -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated by&#160;<a href="https://www.doxygen.org/index.html"><img class="footer" src="doxygen.svg" width="104" height="31" alt="doxygen"/></a> 1.12.0
</small></address>
</div><!-- doc-content -->
</body>
</html>
