<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Read genomic data files (VCF, BCF, BGEN, H5AD) into R/Rcpp in chunks • GenomicDataStream</title>
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="deps/headroom-0.11.0/headroom.min.js"></script><script src="deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="deps/search-1.0.0/fuse.min.js"></script><script src="deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="Read genomic data files (VCF, BCF, BGEN, H5AD) into R/Rcpp in chunks">
<meta name="description" content="Read genomic data files (VCF, BCF, BGEN, H5AD) into R/Rcpp in chunks for analysis with Armadillo or Eigen libraries. BGEN support based on rbgen_v1.1.5.">
<meta property="og:description" content="Read genomic data files (VCF, BCF, BGEN, H5AD) into R/Rcpp in chunks for analysis with Armadillo or Eigen libraries. BGEN support based on rbgen_v1.1.5.">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="index.html">GenomicDataStream</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.5</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/GabrielHoffman/GenomicDataStream/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level2">
<h2 id="genomicdatastream">GenomicDataStream<a class="anchor" aria-label="anchor" href="#genomicdatastream"></a>
</h2>
<div class="section level3">
<h3 id="a-scalable-interface-between-data-and-analysis-underneath-r">A scalable interface between data and analysis underneath R<a class="anchor" aria-label="anchor" href="#a-scalable-interface-between-data-and-analysis-underneath-r"></a>
</h3>
<p><img src="reference/figures/GenomicDataStream.png"></p>
<div align="justify">
<p>Reading genomic data files (<a href="https://www.ebi.ac.uk/training/online/courses/human-genetic-variation-introduction/variant-identification-and-analysis/understanding-vcf-format/" class="external-link">VCF</a>, <a href="https://samtools.github.io/bcftools/howtos/index.html" class="external-link">BCF</a>, <a href="https://www.chg.ox.ac.uk/~gav/bgen_format/index.html" class="external-link">BGEN</a>, <a href="https://anndata.readthedocs.io/en/latest/index.html" class="external-link">H5AD</a>, <a href="https://bioconductor.org/packages/DelayedArray" class="external-link">DelayedArray</a>) into R/Rcpp in chunks for analysis with <nobr><a href="https://doi.org/10.21105/joss.00026" class="external-link">Armadillo</a></nobr> / <a href="eigen.tuxfamily.org">Eigen</a> / <a href="https://www.rcpp.org" class="external-link">Rcpp</a> libraries. Mondern datasets are often too big to fit into memory, and many analyses <nobr>operate</nobr> a small chunk features at a time. Yet in practice, many implementations require the whole dataset stored in memory. Others pair an analysis with a specific data format (i.e. regresson analysis paired with genotype data from a VCF) in way that the two components can’t be separated for use in other applications.</p>
<p>The <code>GenomicDataStream</code> interface separate:</p>
<ol style="list-style-type: decimal">
<li>data source</li>
<li>streaming chunks of features into a data matrix</li>
<li>downstream analysis</li>
</ol>
<p><code>GenomicDataStream</code> provides interfaces at both the C++ and R levels. The C++ interface prioritizes efficiency, while the R interface wraps the C++ backend for non-technical users.</p>
</div>
</div>
<div class="section level3">
<h3 id="see-header-only-c-library-documentation">See header-only C++ library <a href="doxygen/html/index.html">documentation</a>
<a class="anchor" aria-label="anchor" href="#see-header-only-c-library-documentation"></a>
</h3>
</div>
</div>
<div class="section level2">
<h2 id="usage">Usage<a class="anchor" aria-label="anchor" href="#usage"></a>
</h2>
<p>In general, variants from genetic data are used as covariates in <code><a href="https://rdrr.io/pkg/fastlmm/man/lmFitFeatures.html" class="external-link">lmFitFeatures()</a></code>, and genes from single cell data are used as responses in <code><a href="reference/lmFitResponses.html">lmFitResponses()</a></code>.</p>
<div class="section level3">
<h3 id="example-code-with-r">Example code with R<a class="anchor" aria-label="anchor" href="#example-code-with-r"></a>
</h3>
<div class="section level4">
<h4 id="read-genotype-data-into-r">Read genotype data into R<a class="anchor" aria-label="anchor" href="#read-genotype-data-into-r"></a>
</h4>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/GabrielHoffman/GenomicDataStream" class="external-link">GenomicDataStream</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># VCF file</span></span>
<span><span class="va">file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"test.vcf.gz"</span>, package <span class="op">=</span> <span class="st">"GenomicDataStream"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># initialize </span></span>
<span><span class="va">gds</span> <span class="op">=</span> <span class="fu"><a href="reference/GenomicDataStream.html">GenomicDataStream</a></span><span class="op">(</span><span class="va">file</span>, <span class="st">"DS"</span>, chunkSize<span class="op">=</span><span class="fl">5</span>, initialize<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">n</span> <span class="op">=</span> <span class="fl">60</span></span>
<span><span class="va">y</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span><span class="va">n</span><span class="op">)</span></span>
<span><span class="va">design</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fl">1</span>, <span class="va">n</span>, <span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># loop until break</span></span>
<span><span class="kw">while</span><span class="op">(</span> <span class="fl">1</span> <span class="op">)</span><span class="op">{</span></span>
<span></span>
<span>    <span class="co"># get data chunk</span></span>
<span>    <span class="co"># data$X matrix with features as columns</span></span>
<span>    <span class="co"># data$info information about each feature as rows</span></span>
<span>    <span class="va">dat</span> <span class="op">=</span> <span class="fu"><a href="reference/getNextChunk.html">getNextChunk</a></span><span class="op">(</span><span class="va">gds</span><span class="op">)</span></span>
<span></span>
<span>    <span class="co"># check if end of stream </span></span>
<span>    <span class="kw">if</span><span class="op">(</span> <span class="fu"><a href="reference/atEndOfStream.html">atEndOfStream</a></span><span class="op">(</span><span class="va">gds</span><span class="op">)</span> <span class="op">)</span> <span class="kw">break</span></span>
<span>    </span>
<span>    <span class="co"># do analysis on this chunk of data</span></span>
<span>    <span class="va">fit</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/fastlmm/man/lmFitFeatures.html" class="external-link">lmFitFeatures</a></span><span class="op">(</span><span class="va">y</span>, <span class="va">design</span>, <span class="va">dat</span><span class="op">$</span><span class="va">X</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="use-r-to-run-analysis-at-c-level">Use R to run analysis at C++ level<a class="anchor" aria-label="anchor" href="#use-r-to-run-analysis-at-c-level"></a>
</h4>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/GabrielHoffman/GenomicDataStream" class="external-link">GenomicDataStream</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># VCF file</span></span>
<span><span class="va">file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"test.vcf.gz"</span>, package <span class="op">=</span> <span class="st">"GenomicDataStream"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># create object, but don't read yet </span></span>
<span><span class="va">gds</span> <span class="op">=</span> <span class="fu"><a href="reference/GenomicDataStream.html">GenomicDataStream</a></span><span class="op">(</span><span class="va">file</span>, <span class="st">"DS"</span>, chunkSize<span class="op">=</span><span class="fl">5</span><span class="op">)</span></span>
<span></span>
<span><span class="va">n</span> <span class="op">=</span> <span class="fl">60</span></span>
<span><span class="va">y</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span><span class="va">n</span><span class="op">)</span></span>
<span><span class="va">design</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fl">1</span>, <span class="va">n</span>, <span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># regression of y ~ design + X[,j]</span></span>
<span><span class="co">#   where X[,j] is the jth variant in the GenomicDataStream</span></span>
<span><span class="co"># data in GenomicDataStream is only accessed at C++ level </span></span>
<span><span class="va">fit</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/fastlmm/man/lmFitFeatures.html" class="external-link">lmFitFeatures</a></span><span class="op">(</span><span class="va">y</span>, <span class="va">design</span>, <span class="va">gds</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="example-code-with-c17">Example code with C++17<a class="anchor" aria-label="anchor" href="#example-code-with-c17"></a>
</h3>
<div class="sourceCode" id="cb3"><pre class="sourceCode cpp"><code class="sourceCode cpp"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="pp">#include </span><span class="im">&lt;RcppArmadillo.h&gt;</span></span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a><span class="pp">#include </span><span class="im">&lt;GenomicDataStream.h&gt;</span></span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a></span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a><span class="co">// use namespace for GenomicDataStream</span></span>
<span id="cb3-5"><a href="#cb3-5" tabindex="-1"></a><span class="kw">using</span> <span class="kw">namespace</span> gds<span class="op">;</span></span>
<span id="cb3-6"><a href="#cb3-6" tabindex="-1"></a></span>
<span id="cb3-7"><a href="#cb3-7" tabindex="-1"></a><span class="co">// parameters </span></span>
<span id="cb3-8"><a href="#cb3-8" tabindex="-1"></a>string file <span class="op">=</span> <span class="st">"test.vcf.gz"</span><span class="op">;</span></span>
<span id="cb3-9"><a href="#cb3-9" tabindex="-1"></a>string field <span class="op">=</span> <span class="st">"DS"</span><span class="op">;</span>    <span class="co">// read dosage field</span></span>
<span id="cb3-10"><a href="#cb3-10" tabindex="-1"></a>string region <span class="op">=</span> <span class="st">""</span><span class="op">;</span>     <span class="co">// no region filter</span></span>
<span id="cb3-11"><a href="#cb3-11" tabindex="-1"></a>string samples <span class="op">=</span> <span class="st">"-"</span><span class="op">;</span>   <span class="co">// no samples filter</span></span>
<span id="cb3-12"><a href="#cb3-12" tabindex="-1"></a><span class="dt">int</span> chunkSize <span class="op">=</span> <span class="dv">4</span><span class="op">;</span>      <span class="co">// each chunk will read 4 variants</span></span>
<span id="cb3-13"><a href="#cb3-13" tabindex="-1"></a></span>
<span id="cb3-14"><a href="#cb3-14" tabindex="-1"></a><span class="co">// initialize parameters</span></span>
<span id="cb3-15"><a href="#cb3-15" tabindex="-1"></a>Param param<span class="op">(</span>file<span class="op">,</span> region<span class="op">,</span> samples<span class="op">,</span> chunkSize<span class="op">);</span></span>
<span id="cb3-16"><a href="#cb3-16" tabindex="-1"></a>param<span class="op">.</span>setField<span class="op">(</span> field <span class="op">);</span></span>
<span id="cb3-17"><a href="#cb3-17" tabindex="-1"></a></span>
<span id="cb3-18"><a href="#cb3-18" tabindex="-1"></a><span class="co">// Initialise GenomicDataStream to read </span></span>
<span id="cb3-19"><a href="#cb3-19" tabindex="-1"></a><span class="co">// VCF/BCF and BGEN with same interface</span></span>
<span id="cb3-20"><a href="#cb3-20" tabindex="-1"></a>unique_ptr<span class="op">&lt;</span>GenomicDataStream<span class="op">&gt;</span> gdsStream <span class="op">=</span> createFileView<span class="op">(</span> param <span class="op">);</span></span>
<span id="cb3-21"><a href="#cb3-21" tabindex="-1"></a></span>
<span id="cb3-22"><a href="#cb3-22" tabindex="-1"></a><span class="co">// declare DataChunk storing an Armadillo matrix for each chunk</span></span>
<span id="cb3-23"><a href="#cb3-23" tabindex="-1"></a>DataChunk<span class="op">&lt;</span>arma<span class="op">::</span>mat<span class="op">&gt;</span> chunk<span class="op">;</span></span>
<span id="cb3-24"><a href="#cb3-24" tabindex="-1"></a></span>
<span id="cb3-25"><a href="#cb3-25" tabindex="-1"></a><span class="co">// Store meta-data about each variant</span></span>
<span id="cb3-26"><a href="#cb3-26" tabindex="-1"></a>VariantInfo <span class="op">*</span>info<span class="op">;</span></span>
<span id="cb3-27"><a href="#cb3-27" tabindex="-1"></a></span>
<span id="cb3-28"><a href="#cb3-28" tabindex="-1"></a><span class="co">// loop through chunks</span></span>
<span id="cb3-29"><a href="#cb3-29" tabindex="-1"></a><span class="cf">while</span><span class="op">(</span> gdsStream<span class="op">-&gt;</span>getNextChunk<span class="op">(</span> chunk <span class="op">)</span> <span class="op">){</span></span>
<span id="cb3-30"><a href="#cb3-30" tabindex="-1"></a></span>
<span id="cb3-31"><a href="#cb3-31" tabindex="-1"></a>    <span class="co">// get data from chunk</span></span>
<span id="cb3-32"><a href="#cb3-32" tabindex="-1"></a>    <span class="co">// chunk.getData();</span></span>
<span id="cb3-33"><a href="#cb3-33" tabindex="-1"></a></span>
<span id="cb3-34"><a href="#cb3-34" tabindex="-1"></a>    <span class="co">// get variant information</span></span>
<span id="cb3-35"><a href="#cb3-35" tabindex="-1"></a>    info <span class="op">=</span> chunk<span class="op">.</span>getInfo<span class="op">&lt;</span>VariantInfo<span class="op">&gt;();</span></span>
<span id="cb3-36"><a href="#cb3-36" tabindex="-1"></a></span>
<span id="cb3-37"><a href="#cb3-37" tabindex="-1"></a>    <span class="co">// Do analysis with variants in this chunk</span></span>
<span id="cb3-38"><a href="#cb3-38" tabindex="-1"></a><span class="op">}</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="supported-formats">Supported formats<a class="anchor" aria-label="anchor" href="#supported-formats"></a>
</h2>
<div class="section level4">
<h4 id="genetic-data">Genetic data<a class="anchor" aria-label="anchor" href="#genetic-data"></a>
</h4>
<table class="table">
<colgroup>
<col width="14%">
<col width="21%">
<col width="64%">
</colgroup>
<thead><tr>
<th>Format</th>
<th>Version</th>
<th>Support</th>
</tr></thead>
<tbody>
<tr>
<td>BGEN</td>
<td>1.1</td>
<td>biallelic variants</td>
</tr>
<tr>
<td>BGEN</td>
<td>1.2, 1.3</td>
<td>phased or unphased biallelic variants</td>
</tr>
<tr>
<td>PGEN</td>
<td>plink2</td>
<td></td>
</tr>
<tr>
<td>BED</td>
<td>plink1</td>
<td></td>
</tr>
<tr>
<td>VCF / BCF</td>
<td>4.x</td>
<td>biallelic variants with <code>GT/GP</code> fields, continuous dosage with <code>DS</code> field</td>
</tr>
</tbody>
</table>
</div>
<div class="section level4">
<h4 id="single-cell-data">Single cell data<a class="anchor" aria-label="anchor" href="#single-cell-data"></a>
</h4>
<p>Count matrices for single cell data are stored in the H5AD format. This format, based on <a href="https://en.wikipedia.org/wiki/Hierarchical_Data_Format" class="external-link">HDF5</a>, can store millions of cells since it is designed for sparse counts (i.e. many entries are 0) and uses built-in compression. H5AD enables file-backed random access for analyzing a subset of the data without reading the entire file in to memory.</p>
</div>
</div>
<div class="section level2">
<h2 id="dependencies">Dependencies<a class="anchor" aria-label="anchor" href="#dependencies"></a>
</h2>
<table class="table">
<colgroup>
<col width="7%">
<col width="23%">
<col width="69%">
</colgroup>
<thead><tr>
<th>Package</th>
<th>Ref</th>
<th>Role</th>
</tr></thead>
<tbody>
<tr>
<td><a href="https://cran.r-project.org/package=vcfppR" class="external-link">vcfppR</a></td>
<td><a href="https://doi.org/10.1093/bioinformatics/btae049" class="external-link">Bioinformatics</a></td>
<td>C++ API for htslib</td>
</tr>
<tr>
<td><a href="https://github.com/samtools/htslib" class="external-link">htslib</a></td>
<td><a href="https://doi.org/10.1093/gigascience/giab007" class="external-link">GigaScience</a></td>
<td>C API for VCF/BCF files</td>
</tr>
<tr>
<td><a href="https://bioconductor.org/packages/beachmat/" class="external-link">beatchmat</a></td>
<td><a href="https://doi.org/10.1371/journal.pcbi.1006135" class="external-link">PLoS Comp Biol</a></td>
<td>C++ API for access data owned by R</td>
</tr>
<tr>
<td><a href="https://bioconductor.org/packages/DelayedArray/" class="external-link">DelayedArray</a></td>
<td></td>
<td>R interface for handling on-disk data formats</td>
</tr>
<tr>
<td><a href="https://cran.r-project.org/package=Rcpp" class="external-link">Rcpp</a></td>
<td><a href="https://doi.org/10.18637/jss.v040.i08" class="external-link">J Stat Software</a></td>
<td>API for R/C++ integration</td>
</tr>
<tr>
<td><a href="https://cran.r-project.org/package=RcppEigen" class="external-link">RcppEigen</a></td>
<td><a href="https://doi.org/10.18637/jss.v052.i05" class="external-link">J Stat Software</a></td>
<td>API for Rcpp access to Eigen matrix library</td>
</tr>
<tr>
<td><a href="https://cran.r-project.org/package=RcppArmadillo" class="external-link">RcppArmadillo</a></td>
<td><a href="https://doi.org/10.18637/jss.v040.i08" class="external-link">J Stat Software</a></td>
<td>API for Rcpp access to Armadillo matrix library</td>
</tr>
<tr>
<td><a href="https://eigen.tuxfamily.org" class="external-link">Eigen</a></td>
<td></td>
<td>C++ library for linear algebra with advanced features</td>
</tr>
<tr>
<td><a href="https://arma.sourceforge.net" class="external-link">Armadillo</a></td>
<td><a href="https://doi.org/10.21105/joss.00026" class="external-link">J Open Src Soft</a></td>
<td>User-friendly C++ library for linear algebra</td>
</tr>
</tbody>
</table>
</div>
<div class="section level2">
<h2 id="notes">Notes<a class="anchor" aria-label="anchor" href="#notes"></a>
</h2>
<p><code>GenomicDataStream</code> provide flexability in terms of data input types and and matrix libraries. This can useful in many cases, but the large number of dependencies can require installation of additional libraries and increase compile times. Some of these dependencies can be avoided by removing support for some capabilities with compiler flags in <code>Makevars</code>:</p>
<p><code>-D DISABLE_DELAYED_STREAM</code><br>
           Omit <code>DelayedStream</code> class, remove dependence on <code>Rcpp</code> and <code>beachmat</code></p>
<p><code>-D DISABLE_EIGEN</code><br>
           Omit support for Eigen matrix library, and remove dependence on <code>RcppEigen</code> and <code>Eigen</code></p>
<p><code>-D DISABLE_RCPP</code><br>
           Omit support for <code>Rcpp</code> matrix library, and remove dependence on <code>Rcpp</code></p>
<p><code>GenomicDataStream</code> is written so that core functions are in C++17 with no dependence or R or Rcpp. On top of that, there is a thin wrapper that uses Rcpp to interface between R and the lower-level library.</p>
</div>

  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/GabrielHoffman/GenomicDataStream/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/GabrielHoffman/GenomicDataStream/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/Artistic-2.0" class="external-link">Artistic-2.0</a></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing GenomicDataStream</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>
<a href="http://gabrielhoffman.github.io" class="external-link">Gabriel Hoffman</a> <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0002-0957-0224" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
</ul>
</div>



  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="http://gabrielhoffman.github.io" class="external-link">Gabriel Hoffman</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
