<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Data as a GenomicDataStream • GenomicDataStream</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Data as a GenomicDataStream">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">GenomicDataStream</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.10</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item"><a class="nav-link" href="../articles/GenomicDataStream.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/extensions.html">Extending GenomicDataStream</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/GabrielHoffman/GenomicDataStream/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Data as a GenomicDataStream</h1>
            <h3 data-toc-skip class="subtitle">R and C++ code</h3>
                        <h4 data-toc-skip class="author">Developed by <a href="http://gabrielhoffman.github.io/" class="external-link">Gabriel Hoffman</a>
</h4>
            
            <h4 data-toc-skip class="date">Run on 2025-01-23
15:22:14</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/GabrielHoffman/GenomicDataStream/blob/HEAD/vignettes/GenomicDataStream.Rmd" class="external-link"><code>vignettes/GenomicDataStream.Rmd</code></a></small>
      <div class="d-none name"><code>GenomicDataStream.Rmd</code></div>
    </div>

    
    
<!---


--->
<style>
body {
text-align: justify}
</style>
<p><code>GenomicDataStream</code> designed to chunks of
<em>features</em> rather than chunks of <em>samples</em>. Features are
stored as columns in the matrix returned by R/C++, <ins>independent of
the underlying data storage format</ins>.</p>
<div class="section level2">
<h2 id="usage">Usage<a class="anchor" aria-label="anchor" href="#usage"></a>
</h2>
<p>In general, variants from genetic data are used as covariates in
<code>lmFitFeatures()</code>, and genes from single cell data are used
as responses in <code>lmFitResponses()</code>.</p>
<div class="section level4">
<h4 id="example-code-with-r">Example code with R<a class="anchor" aria-label="anchor" href="#example-code-with-r"></a>
</h4>
<div class="section level5">
<h5 id="read-genotype-data-into-r">Read genotype data into R<a class="anchor" aria-label="anchor" href="#read-genotype-data-into-r"></a>
</h5>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/GabrielHoffman/GenomicDataStream" class="external-link">GenomicDataStream</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># VCF file</span></span>
<span><span class="va">file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"test.vcf.gz"</span>, package <span class="op">=</span> <span class="st">"GenomicDataStream"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># initialize </span></span>
<span><span class="va">gds</span> <span class="op">=</span> <span class="fu"><a href="../reference/GenomicDataStream.html">GenomicDataStream</a></span><span class="op">(</span><span class="va">file</span>, <span class="st">"DS"</span>, chunkSize<span class="op">=</span><span class="fl">5</span>, initialize<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">n</span> <span class="op">=</span> <span class="fl">60</span></span>
<span><span class="va">y</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span><span class="va">n</span><span class="op">)</span></span>
<span><span class="va">design</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fl">1</span>, <span class="va">n</span>, <span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># loop until break</span></span>
<span><span class="kw">while</span><span class="op">(</span> <span class="fl">1</span> <span class="op">)</span><span class="op">{</span></span>
<span></span>
<span>    <span class="co"># get data chunk</span></span>
<span>    <span class="co"># data$X matrix with features as columns</span></span>
<span>    <span class="co"># data$info information about each feature as rows</span></span>
<span>    <span class="va">dat</span> <span class="op">=</span> <span class="fu"><a href="../reference/getNextChunk.html">getNextChunk</a></span><span class="op">(</span><span class="va">gds</span><span class="op">)</span></span>
<span></span>
<span>    <span class="co"># check if end of stream </span></span>
<span>    <span class="kw">if</span><span class="op">(</span> <span class="fu"><a href="../reference/atEndOfStream.html">atEndOfStream</a></span><span class="op">(</span><span class="va">gds</span><span class="op">)</span> <span class="op">)</span> <span class="kw">break</span></span>
<span>    </span>
<span>    <span class="co"># do analysis on this chunk of data</span></span>
<span>    <span class="va">fit</span> <span class="op">=</span> <span class="fu">lmFitFeatures</span><span class="op">(</span><span class="va">y</span>, <span class="va">design</span>, <span class="va">dat</span><span class="op">$</span><span class="va">X</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
</div>
<div class="section level5">
<h5 id="use-r-to-run-analysis-at-c-level">Use R to run analysis at C++ level<a class="anchor" aria-label="anchor" href="#use-r-to-run-analysis-at-c-level"></a>
</h5>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/GabrielHoffman/GenomicDataStream" class="external-link">GenomicDataStream</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># VCF file</span></span>
<span><span class="va">file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"test.vcf.gz"</span>, package <span class="op">=</span> <span class="st">"GenomicDataStream"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># create object, but don't read yet </span></span>
<span><span class="co"># Read DS field storing dosage</span></span>
<span><span class="va">gds</span> <span class="op">=</span> <span class="fu"><a href="../reference/GenomicDataStream.html">GenomicDataStream</a></span><span class="op">(</span><span class="va">file</span>, <span class="st">"DS"</span>, chunkSize<span class="op">=</span><span class="fl">5</span><span class="op">)</span></span>
<span></span>
<span><span class="va">n</span> <span class="op">=</span> <span class="fl">60</span></span>
<span><span class="va">y</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span><span class="va">n</span><span class="op">)</span></span>
<span><span class="va">design</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fl">1</span>, <span class="va">n</span>, <span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># regression of y ~ design + X[,j]</span></span>
<span><span class="co">#   where X[,j] is the jth variant in the GenomicDataStream</span></span>
<span><span class="co"># data in GenomicDataStream is only accessed at C++ level </span></span>
<span><span class="va">fit</span> <span class="op">=</span> <span class="fu">lmFitFeatures</span><span class="op">(</span><span class="va">y</span>, <span class="va">design</span>, <span class="va">gds</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level4">
<h4 id="example-code-with-c17">Example code with C++17<a class="anchor" aria-label="anchor" href="#example-code-with-c17"></a>
</h4>
<div class="sourceCode" id="cb3"><pre class="sourceCode cpp"><code class="sourceCode cpp"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="pp">#include </span><span class="im">&lt;RcppArmadillo.h&gt;</span></span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a><span class="pp">#include </span><span class="im">&lt;GenomicDataStream.h&gt;</span></span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a></span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a><span class="co">// use namespace for GenomicDataStream</span></span>
<span id="cb3-5"><a href="#cb3-5" tabindex="-1"></a><span class="kw">using</span> <span class="kw">namespace</span> gds<span class="op">;</span></span>
<span id="cb3-6"><a href="#cb3-6" tabindex="-1"></a></span>
<span id="cb3-7"><a href="#cb3-7" tabindex="-1"></a><span class="co">// parameters </span></span>
<span id="cb3-8"><a href="#cb3-8" tabindex="-1"></a>string file <span class="op">=</span> <span class="st">"test.vcf.gz"</span><span class="op">;</span></span>
<span id="cb3-9"><a href="#cb3-9" tabindex="-1"></a>string field <span class="op">=</span> <span class="st">"DS"</span><span class="op">;</span>    <span class="co">// read dosage field</span></span>
<span id="cb3-10"><a href="#cb3-10" tabindex="-1"></a>string region <span class="op">=</span> <span class="st">""</span><span class="op">;</span>     <span class="co">// no region filter</span></span>
<span id="cb3-11"><a href="#cb3-11" tabindex="-1"></a>string samples <span class="op">=</span> <span class="st">"-"</span><span class="op">;</span>   <span class="co">// no samples filter</span></span>
<span id="cb3-12"><a href="#cb3-12" tabindex="-1"></a><span class="dt">int</span> chunkSize <span class="op">=</span> <span class="dv">4</span><span class="op">;</span>      <span class="co">// each chunk will read 4 variants</span></span>
<span id="cb3-13"><a href="#cb3-13" tabindex="-1"></a></span>
<span id="cb3-14"><a href="#cb3-14" tabindex="-1"></a><span class="co">// initialize parameters</span></span>
<span id="cb3-15"><a href="#cb3-15" tabindex="-1"></a>Param param<span class="op">(</span>file<span class="op">,</span> region<span class="op">,</span> samples<span class="op">,</span> chunkSize<span class="op">);</span></span>
<span id="cb3-16"><a href="#cb3-16" tabindex="-1"></a>param<span class="op">.</span>setField<span class="op">(</span> field <span class="op">);</span></span>
<span id="cb3-17"><a href="#cb3-17" tabindex="-1"></a></span>
<span id="cb3-18"><a href="#cb3-18" tabindex="-1"></a><span class="co">// Initialise GenomicDataStream to read </span></span>
<span id="cb3-19"><a href="#cb3-19" tabindex="-1"></a><span class="co">// VCF/BCF and BGEN with same interface</span></span>
<span id="cb3-20"><a href="#cb3-20" tabindex="-1"></a>unique_ptr<span class="op">&lt;</span>GenomicDataStream<span class="op">&gt;</span> gdsStream <span class="op">=</span> createFileView<span class="op">(</span> param <span class="op">);</span></span>
<span id="cb3-21"><a href="#cb3-21" tabindex="-1"></a></span>
<span id="cb3-22"><a href="#cb3-22" tabindex="-1"></a><span class="co">// declare DataChunk storing an Armadillo matrix for each chunk</span></span>
<span id="cb3-23"><a href="#cb3-23" tabindex="-1"></a>DataChunk<span class="op">&lt;</span>arma<span class="op">::</span>mat<span class="op">&gt;</span> chunk<span class="op">;</span></span>
<span id="cb3-24"><a href="#cb3-24" tabindex="-1"></a></span>
<span id="cb3-25"><a href="#cb3-25" tabindex="-1"></a><span class="co">// Store meta-data about each variant</span></span>
<span id="cb3-26"><a href="#cb3-26" tabindex="-1"></a>VariantInfo <span class="op">*</span>info<span class="op">;</span></span>
<span id="cb3-27"><a href="#cb3-27" tabindex="-1"></a></span>
<span id="cb3-28"><a href="#cb3-28" tabindex="-1"></a><span class="co">// loop through chunks</span></span>
<span id="cb3-29"><a href="#cb3-29" tabindex="-1"></a><span class="cf">while</span><span class="op">(</span> gdsStream<span class="op">-&gt;</span>getNextChunk<span class="op">(</span> chunk <span class="op">)</span> <span class="op">){</span></span>
<span id="cb3-30"><a href="#cb3-30" tabindex="-1"></a></span>
<span id="cb3-31"><a href="#cb3-31" tabindex="-1"></a>    <span class="co">// get data from chunk</span></span>
<span id="cb3-32"><a href="#cb3-32" tabindex="-1"></a>    <span class="co">// chunk.getData();</span></span>
<span id="cb3-33"><a href="#cb3-33" tabindex="-1"></a></span>
<span id="cb3-34"><a href="#cb3-34" tabindex="-1"></a>    <span class="co">// get variant information</span></span>
<span id="cb3-35"><a href="#cb3-35" tabindex="-1"></a>    info <span class="op">=</span> chunk<span class="op">.</span>getInfo<span class="op">&lt;</span>VariantInfo<span class="op">&gt;();</span></span>
<span id="cb3-36"><a href="#cb3-36" tabindex="-1"></a></span>
<span id="cb3-37"><a href="#cb3-37" tabindex="-1"></a>    <span class="co">// Do analysis with variants in this chunk</span></span>
<span id="cb3-38"><a href="#cb3-38" tabindex="-1"></a><span class="op">}</span></span></code></pre></div>
<p>Accessing data</p>
</div>
</div>
<div class="section level2">
<h2 id="data-types">Data types<a class="anchor" aria-label="anchor" href="#data-types"></a>
</h2>
<p>At the C++ level data from the <code>GenomicDataStream</code> can be
accessed as multiple types a according to the templated definition of
<code>DataChunk</code>. Above, each <code>chunk</code> is a
<code>DataChunk&lt;arma::mat&gt;</code> so the data is returned as an
Armadillo matrix of doubles (i.e. <code>arma::mat</code>).</p>
<p>Suppported data types for dense matricies are:</p>
<table class="table">
<thead><tr class="header">
<th>Library</th>
<th>Namespace</th>
<th>Type</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Rcpp</td>
<td><code>Rcpp</code></td>
<td><code>NumericMatrix</code></td>
</tr>
<tr class="even">
<td>Armadillo</td>
<td><code>arma</code></td>
<td><code>mat</code></td>
</tr>
<tr class="odd">
<td>Eigen</td>
<td><code>Eigen</code></td>
<td><code>MatrixXd</code></td>
</tr>
<tr class="even">
<td>STL</td>
<td><code>std</code></td>
<td><code>vector&lt;double&gt;</code></td>
</tr>
</tbody>
</table>
<p>Importantly, each of these types wraps an array of
<code>double</code> storing data in column-major order. The types just
provide different interfaces to the underlying array for use in
downstream analysis. In each case, the raw data is read as an
<code>double</code> array, a constructor is called for the requested
data type using the array as the underlying data and an object is
returned to the user. In fact, the constructors to the dense matrix
types for Eigen and Armadillo return objects that point to the original
<code>double</code> array, without allocating new memory.</p>
<p><code>GenomicDataStream</code> also supports the following sparse
matrix types:</p>
<table class="table">
<thead><tr class="header">
<th>Library</th>
<th>Namespace</th>
<th>Type</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Armadillo</td>
<td><code>arma</code></td>
<td><code>sp_mat</code></td>
</tr>
<tr class="even">
<td>Eigen</td>
<td><code>Eigen</code></td>
<td><code>SparseMatrix</code></td>
</tr>
</tbody>
</table>
<p>In each case, a constructor is called for the requested data type the
converter <code>double</code> array to a sparse matrix.</p>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<details><pre><code><span><span class="co">## R version 4.4.1 (2024-06-14)</span></span>
<span><span class="co">## Platform: aarch64-apple-darwin23.5.0</span></span>
<span><span class="co">## Running under: macOS Sonoma 14.7.1</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /Users/gabrielhoffman/prog/R-4.4.1/lib/libRblas.dylib </span></span>
<span><span class="co">## LAPACK: /opt/homebrew/Cellar/r/4.4.2_2/lib/R/lib/libRlapack.dylib;  LAPACK version 3.12.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: America/New_York</span></span>
<span><span class="co">## tzcode source: internal</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">## [1] readr_2.1.5</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##  [1] vctrs_0.6.5       cli_3.6.3         knitr_1.49        rlang_1.1.4       xfun_0.49        </span></span>
<span><span class="co">##  [6] textshaping_0.4.0 jsonlite_1.8.9    glue_1.8.0        htmltools_0.5.8.1 ragg_1.3.3       </span></span>
<span><span class="co">## [11] sass_0.4.9        hms_1.1.3         fansi_1.0.6       rmarkdown_2.29    evaluate_1.0.1   </span></span>
<span><span class="co">## [16] jquerylib_0.1.4   tibble_3.2.1      tzdb_0.4.0        fastmap_1.2.0     yaml_2.3.10      </span></span>
<span><span class="co">## [21] lifecycle_1.0.4   compiler_4.4.1    codetools_0.2-20  fs_1.6.5          htmlwidgets_1.6.4</span></span>
<span><span class="co">## [26] pkgconfig_2.0.3   systemfonts_1.1.0 digest_0.6.37     R6_2.5.1          utf8_1.2.4       </span></span>
<span><span class="co">## [31] pillar_1.9.0      magrittr_2.0.3    bslib_0.8.0       tools_4.4.1       pkgdown_2.1.1    </span></span>
<span><span class="co">## [36] cachem_1.1.0      desc_1.4.3</span></span></code></pre>
<p>&lt;&gt;</p>
<!-- # References -->
</details>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="http://gabrielhoffman.github.io" class="external-link">Gabriel Hoffman</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
