<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>GenomicDataStream — GenomicDataStream • GenomicDataStream</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="GenomicDataStream — GenomicDataStream"><meta name="description" content="Read genomic data files (VCF, BCF, BGEN, h5ad) into R/Rcpp in chunks for analysis with Armadillo or Eigen libraries
Interface to GenomicDataStream C++ code"><meta property="og:description" content="Read genomic data files (VCF, BCF, BGEN, h5ad) into R/Rcpp in chunks for analysis with Armadillo or Eigen libraries
Interface to GenomicDataStream C++ code"><meta property="og:image" content="http://gabrielhoffman.github.io/GenomicDataStream/logo.svg"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">GenomicDataStream</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.16</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/GenomicDataStream.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/extensions.html">Extending GenomicDataStream</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/GabrielHoffman/GenomicDataStream/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.svg" class="logo" alt=""><h1>GenomicDataStream</h1>
      <small class="dont-index">Source: <a href="https://github.com/GabrielHoffman/GenomicDataStream/blob/HEAD/R/GenomicDataStream.R" class="external-link"><code>R/GenomicDataStream.R</code></a></small>
      <div class="d-none name"><code>GenomicDataStream.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Read genomic data files (VCF, BCF, BGEN, h5ad) into R/Rcpp in chunks for analysis with Armadillo or Eigen libraries</p>
<p>Interface to GenomicDataStream C++ code</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">GenomicDataStream</span><span class="op">(</span></span>
<span>  <span class="va">file</span>,</span>
<span>  field <span class="op">=</span> <span class="st">""</span>,</span>
<span>  region <span class="op">=</span> <span class="st">""</span>,</span>
<span>  samples <span class="op">=</span> <span class="st">"-"</span>,</span>
<span>  MAF <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  minVariance <span class="op">=</span> <span class="fl">2</span> <span class="op">*</span> <span class="op">(</span><span class="fl">1</span> <span class="op">-</span> <span class="va">MAF</span><span class="op">)</span> <span class="op">*</span> <span class="va">MAF</span>,</span>
<span>  chunkSize <span class="op">=</span> <span class="fl">10000</span>,</span>
<span>  missingToMean <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  initialize <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-file">file<a class="anchor" aria-label="anchor" href="#arg-file"></a></dt>
<dd><p>file in VCF/BCF/BGEN/PGEN format with index</p></dd>


<dt id="arg-field">field<a class="anchor" aria-label="anchor" href="#arg-field"></a></dt>
<dd><p>field of VCF/BCF to read</p></dd>


<dt id="arg-region">region<a class="anchor" aria-label="anchor" href="#arg-region"></a></dt>
<dd><p>target in the format <code>chr2:1-12345</code>. Multiple regions can be separated by one of <code>",\n\t"</code>, for example <code>"chr2:1-12345, chr3:1000-8000"</code>. Setting region to <code>""</code> includes all variants</p></dd>


<dt id="arg-samples">samples<a class="anchor" aria-label="anchor" href="#arg-samples"></a></dt>
<dd><p>string of comma separated sample IDs to extract: <code>"ID1,ID2,ID3"</code>.  <code>"-"</code> indicates all samples</p></dd>


<dt id="arg-maf">MAF<a class="anchor" aria-label="anchor" href="#arg-maf"></a></dt>
<dd><p>generalized minor allele frequency cutoff keeps variants with variance &gt; \(2(1-f)f\).</p></dd>


<dt id="arg-minvariance">minVariance<a class="anchor" aria-label="anchor" href="#arg-minvariance"></a></dt>
<dd><p>features with variance <code>&gt;= minVariance</code> are retained.  Defaults to \(2(1-f)f\). If <code>NaN</code>, no filtering is applied.</p></dd>


<dt id="arg-chunksize">chunkSize<a class="anchor" aria-label="anchor" href="#arg-chunksize"></a></dt>
<dd><p>number of variants to return per chunk</p></dd>


<dt id="arg-missingtomean">missingToMean<a class="anchor" aria-label="anchor" href="#arg-missingtomean"></a></dt>
<dd><p>if true, set missing values to the mean dosage value. if false, set to <code>NaN</code></p></dd>


<dt id="arg-initialize">initialize<a class="anchor" aria-label="anchor" href="#arg-initialize"></a></dt>
<dd><p>default <code>FALSE</code>.  If <code>TRUE</code>, file info is read from path, otherwise store path until <code>GenomicDataStream</code> is initialized later</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>object of class <code>GenomicDataStream</code></p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>Consider minor allele frequency (MAF) \(f\) and  Hardy-Weinberg equilibrium, the allelic states have probability \(f^2, 2f(1-f), (1-f)^2\). If the variant has mean \(\mu\) and variance \(\sigma^2\), MAF can be estimated from the mean as \(min(\mu/2, 1 - \mu/2)\) or from the variance as \(p = 1+sqrt(1-2\sigma^2))/2\) and MAF = \(min(p, 1-p)\).  In addition the sample variance of the variant is \(2(1-f)f\).  Therefore, setting a MAF cutoff corresponds to a variance cutoff that can also apply to multi-allelic variants.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"test.vcf.gz"</span>, package <span class="op">=</span> <span class="st">"GenomicDataStream"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># initialize</span></span></span>
<span class="r-in"><span><span class="va">obj</span> <span class="op">&lt;-</span> <span class="fu">GenomicDataStream</span><span class="op">(</span><span class="va">file</span>, <span class="st">"DS"</span>, chunkSize <span class="op">=</span> <span class="fl">5</span>, initialize <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">obj</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 		 GenomicDataStream </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   file:          test.vcf.gz </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   initialized:   TRUE </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   stream type:   vcf.gz </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   field:         DS </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   region:         </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   samples:       60 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   minVar cutoff: 0 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   missingToMean: TRUE </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   chunkSize:     5 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   features read: 0 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   end of stream: FALSE </span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># loop until break</span></span></span>
<span class="r-in"><span><span class="kw">while</span> <span class="op">(</span><span class="fl">1</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span>  <span class="co"># get data chunk</span></span></span>
<span class="r-in"><span>  <span class="co"># data$X matrix with features as columns</span></span></span>
<span class="r-in"><span>  <span class="co"># data$info information about each feature as rows</span></span></span>
<span class="r-in"><span>  <span class="va">dat</span> <span class="op">&lt;-</span> <span class="fu"><a href="getNextChunk.html">getNextChunk</a></span><span class="op">(</span><span class="va">obj</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span>  <span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="atEndOfStream.html">atEndOfStream</a></span><span class="op">(</span><span class="va">obj</span><span class="op">)</span><span class="op">)</span> <span class="kw">break</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">dat</span><span class="op">$</span><span class="va">info</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   CHROM   POS          ID A1 A2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1     1 10000 1:10000:C:A  C  A</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2     1 11000 1:11000:T:C  T  C</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3     1 12000 1:12000:T:C  T  C</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4     1 13000 1:13000:T:C  T  C</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5     1 14000 1:14000:G:C  G  C</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   CHROM   POS          ID A1 A2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1     1 15000 1:15000:A:C  A  C</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2     1 16000 1:16000:G:A  G  A</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3     1 17000 1:17000:C:A  C  A</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4     1 18000 1:18000:C:G  C  G</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5     1 19000 1:19000:T:G  T  G</span>
<span class="r-in"><span><span class="co">#</span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="http://gabrielhoffman.github.io" class="external-link">Gabriel Hoffman</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

