<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Package index â€¢ GenomicDataStream</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Package index"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">GenomicDataStream</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.4</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/GabrielHoffman/GenomicDataStream/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-index">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Package index</h1>
    </div>

    <div class="section level2">
      <h2 id="all-functions">All functions<a class="anchor" aria-label="anchor" href="#all-functions"></a></h2>




    </div><div class="section level2">




      <dl><dt>

          <code><a href="AlleleCodeBuf.html">AlleleCodeBuf()</a></code>

        </dt>
        <dd>Returns an empty two-row numeric matrix that ReadAlleles() can load to.</dd>
      </dl><dl><dt>

          <code><a href="BoolBuf.html">BoolBuf()</a></code>

        </dt>
        <dd>Returns a bool buffer that ReadAlleles() can load phasing information to.</dd>
      </dl><dl><dt>

          <code><a href="Buf.html">Buf()</a></code>

        </dt>
        <dd>Returns a numeric buffer that Read() or ReadHardcalls() can load to.</dd>
      </dl><dl><dt>

          <code><a href="ClosePgen.html">ClosePgen()</a></code>

        </dt>
        <dd>Closes a pgen object, releasing resources.</dd>
      </dl><dl><dt>

          <code><a href="ClosePvar.html">ClosePvar()</a></code>

        </dt>
        <dd>Closes a pvar object, releasing memory.</dd>
      </dl><dl><dt>

          <code><a href="GenomicDataStream.html">GenomicDataStream()</a></code>

        </dt>
        <dd>GenomicDataStream</dd>
      </dl><dl><dt>

          <code><a href="GetAlleleCode.html">GetAlleleCode()</a></code>

        </dt>
        <dd>Look up an allele code.</dd>
      </dl><dl><dt>

          <code><a href="GetAlleleCt.html">GetAlleleCt()</a></code>

        </dt>
        <dd>Returns the effective number of alleles for a variant. Note that if no pvar was provided to the NewPgen() call, this function may return 2 even at multiallelic variants, since the .pgen may not store allele-count information.</dd>
      </dl><dl><dt>

          <code><a href="GetMaxAlleleCt.html">GetMaxAlleleCt()</a></code>

        </dt>
        <dd>Returns the maximum GetAlleleCt() value across all variants in the file.</dd>
      </dl><dl><dt>

          <code><a href="GetRawSampleCt.html">GetRawSampleCt()</a></code>

        </dt>
        <dd>Returns the number of samples in the file.</dd>
      </dl><dl><dt>

          <code><a href="GetVariantCt.html">GetVariantCt()</a></code>

        </dt>
        <dd>Returns the number of variants in the file.</dd>
      </dl><dl><dt>

          <code><a href="GetVariantId.html">GetVariantId()</a></code>

        </dt>
        <dd>Convert variant index to variant ID string.</dd>
      </dl><dl><dt>

          <code><a href="GetVariantsById.html">GetVariantsById()</a></code>

        </dt>
        <dd>Convert variant ID string to variant index(es).</dd>
      </dl><dl><dt>

          <code><a href="HardcallPhasePresent.html">HardcallPhasePresent()</a></code>

        </dt>
        <dd>Returns whether explicitly phased hardcalls are present.</dd>
      </dl><dl><dt>

          <code><a href="IntAlleleCodeBuf.html">IntAlleleCodeBuf()</a></code>

        </dt>
        <dd>Returns an empty two-row integer matrix that ReadAlleles() can load to.</dd>
      </dl><dl><dt>

          <code><a href="IntBuf.html">IntBuf()</a></code>

        </dt>
        <dd>Returns an integer buffer that ReadHardcalls() can load to.</dd>
      </dl><dl><dt>

          <code><a href="NewPgen.html">NewPgen()</a></code>

        </dt>
        <dd>Opens a .pgen or PLINK 1 .bed file.</dd>
      </dl><dl><dt>

          <code><a href="NewPvar.html">NewPvar()</a></code>

        </dt>
        <dd>Loads variant IDs and allele codes from a .pvar or .bim file (which can be compressed with gzip or Zstd).</dd>
      </dl><dl><dt>

          <code><a href="Read.html">Read()</a></code>

        </dt>
        <dd>Loads the variant_numth variant, and then fills buf with numeric dosages in [0, 2] indicating the dosages of the first ALT (or user-specified) allele for each sample, with missing values represented by NA.</dd>
      </dl><dl><dt>

          <code><a href="ReadAlleles.html">ReadAlleles()</a></code>

        </dt>
        <dd>Loads the variant_numth variant, and then fills acbuf with integer allele codes, where each column of the buffer corresponds to a sample. An allele code of 0 corresponds to the REF allele, 1 to the first ALT, 2 to the second ALT, etc. Missing hardcalls are represented by a pair of NA codes.</dd>
      </dl><dl><dt>

          <code><a href="ReadHardcalls.html">ReadHardcalls()</a></code>

        </dt>
        <dd>Loads the variant_numth variant, and then fills buf with {0, 1, 2, NA} values indicating the number of copies of the first ALT (or user-specified) allele each sample has.</dd>
      </dl><dl><dt>

          <code><a href="ReadIntList.html">ReadIntList()</a></code>

        </dt>
        <dd>Load hardcalls for multiple variants as an integer matrix.</dd>
      </dl><dl><dt>

          <code><a href="ReadList.html">ReadList()</a></code>

        </dt>
        <dd>Load dosages for multiple variants as a numeric matrix.</dd>
      </dl><dl><dt>

          <code><a href="VariantScores.html">VariantScores()</a></code>

        </dt>
        <dd>Compute variant scores.</dd>
      </dl><dl><dt>

          <code><a href="atEndOfStream.html">atEndOfStream()</a></code>

        </dt>
        <dd>Detected if end of stream is reaached</dd>
      </dl><dl><dt>

          <code><a href="featuresRead.html">featuresRead()</a></code>

        </dt>
        <dd>Get number of features read from enomicDataStream</dd>
      </dl><dl><dt>

          <code><a href="getNextChunk.html">getNextChunk()</a></code>

        </dt>
        <dd>Get data chunk from GenomicDataStream</dd>
      </dl><dl><dt>

          <code><a href="initializeStream.html">initializeStream()</a></code>

        </dt>
        <dd>Initialize GenomicDataStream</dd>
      </dl><dl><dt>

          <code><a href="isInitialized.html">isInitialized()</a></code>

        </dt>
        <dd>Get status of GenomicDataStream</dd>
      </dl><dl><dt>

          <code><a href="print-methods.html">print(<i>&lt;GenomicDataStream&gt;</i>)</a></code>

        </dt>
        <dd>Print object</dd>
      </dl><dl><dt>

          <code><a href="show-methods.html">show(<i>&lt;GenomicDataStream&gt;</i>)</a></code>

        </dt>
        <dd>Show object</dd>
      </dl></div>
  </main></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="http://gabrielhoffman.github.io" class="external-link">Gabriel Hoffman</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

